### DISCUSSION TITLE
How does one find all elements regardless of ifs?

### DISCUSSION THREAD

**Discussion by forthrin** [sparklemotion/nokogiri#2994] [General Q&A] [answered]
```
Nokogiri.HTML5('<!--[if foo]><p>bar</p><![endif]-->').css('body *').each do |e|
  puts e
end
```

Nothing is printed. Suspect if statements are honored, or assumed false.

Removing all comments is not an option because they are needed later.

**Comment by flavorjones** [comment]
@forthrin Thanks for asking this question. I've moved it from the nokogiri.org repository to the nokogiri repository.

Let's look at the DOM that's created:

```ruby
Nokogiri.HTML5('<!--[if foo]><p>bar</p><![endif]-->').to_html
# => "<!--[if foo]><p>bar</p><![endif]--><html><head></head><body></body></html>"

Nokogiri.HTML5('<!--[if foo]><p>bar</p><![endif]-->')
# => 
# #(Document:0x33d9c {
#   name = "document",
#   children = [
#     #(Comment "[if foo]><p>bar</p><![endif]"),
#     #(Element:0x33db0 {
#       name = "html",
#       children = [ #(Element:0x33dc4 { name = "head" }), #(Element:0x33dd8 { name = "body" })]
#       })]
#   })
```

Note that the initial comment is handled while the parser is in the ["initial"](https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode) or ["before html"](https://html.spec.whatwg.org/multipage/parsing.html#the-before-html-insertion-mode) state, and so the comment is inserted into the document before the `<html>` tag.

The `Node#css` will compile the selector `"body *"` to the xpath query `"//body//*"`, which searches for any elements under a "body" element. As you can see from the DOM above, the body of the HTML document is empty, and so the selector correctly does not match any elements.

Please note, that a bare comment is not a well-formed HTML document; you may have better results parsing fragments with the `DocumentFragment` API:

```ruby
Nokogiri::HTML5.fragment('<!--[if foo]><p>bar</p><![endif]-->').to_html
# => "<!--[if foo]><p>bar</p><![endif]-->"

Nokogiri::HTML5.fragment('<!--[if foo]><p>bar</p><![endif]-->')
# => #(DocumentFragment:0x1c034 { name = "#document-fragment", children = [ #(Comment "[if foo]><p>bar</p><![endif]")] })

```

If you have more questions, I can try to answer -- please tell us more about your use case and what you _expected_ to happen.

**Comment by forthrin** [comment]
Thanks for your very kind and helpful reply. To answer the [XY problem](https://en.wikipedia.org/wiki/XY_problem), the use case is:

Remove any possibility of an HTML mail contacting a server except on explicit user interaction (ie. clicking a link).

The current approach is:

1. Parse `*`
    1. Temporarily replace all `<a href>` with a SHA1 of the link (and store in a hash) -- to avoid false positives in safety net
    2. Delete all other attributes containing `://`
    3. Replace all `url(...)` in `<style>` with `none`
1. Safety net: If resulting plaintext HTML contains `://`, throw exception and fix code
1. Restore all `<a href>`
1. Remove `<iframe>`, `<script>`, etc.

So the problem is simply that URLs inside if statements are not removed, and could hence be loaded in a browser.

Any simpler and smarter approach to accomplish the overall goal is appreciated.

**Reply by flavorjones** [reply, marked as answer]
Conditional comments aren't really supported in modern browsers -- https://en.wikipedia.org/wiki/Conditional_comment -- so you may simply want to remove all comments.

But it sounds like what you _really_ want is an HTML sanitizer. I suggest you take a look at https://github.com/flavorjones/loofah or https://github.com/rgrove/sanitize

**Reply by flavorjones** [reply]
Worth noting: Loofah does not sanitize the contents of `<style>` elements, but the Sanitize gem does. Probably worth starting there!
